be1-go/hub/standard_hub/hub_state/queries.go

```diff
@@ -43,12 +49,23 @@ func (q *Queries) GetNextID() int {
 	return id
 }
 
-// SetQueryState sets the state of the query with the given ID
-func (q *Queries) SetQueryState(id int, state bool) {
+// SetQueryReceived sets the state of the query with the given ID as received
+func (q *Queries) SetQueryReceived(id int) error {
 	q.Lock()
 	defer q.Unlock()
 
-	q.state[id] = &state
+	currentState, ok := q.state[id]
+
+	if !ok {
+		return xerrors.Errorf("query with id %d not found", id)
+	}
+
+	if currentState {
+		return xerrors.Errorf("query with id %d already answered", id)
```

Comment:
Just realized something, completely outside the scope of this PR.
This error propagates all the way to the socket.
So if another peer (for whatever reason) sends the "answer" message twice, they will get an error in return saying "we've already gotten your answer".
Is this something that would need to be shared with the other peer, or just logged ? 
What would the other peer be expected to do - retry ? That would likely make things worse (infinite loop of errors / retries)
It seems manageable (in this case) to implement some degree of idempotency.

Reply:
I agree that we should just log it because we wouldn't expect the peer to do anything. I think it could be a nice first issue so I'll add it to the repo. Regarding idempotency, would that also apply to when we try to create a channel that was already created ? Currently we send a "duplicate resource" error but how do we differentiate between cases that warrant idempotency and those that don't ?

Unrelated to that but I also noticed that we don't always respect the [protocol](https://github.com/dedis/popstellar/blob/master/docs/protocol.md#rpc-answer-error) when it comes to sending errors back. Since we have many "levels" of returning in go, the nature of the error gets lost by the time it arrives to the method that's supposed to create the error to send back so we don't have the right code. Thankfully, we already have methods designed to create the right error with the right code but those aren't always used for some reason. I'll investigate the individual cases and open PRs for them. They're easy fixes but also a good way to get to know the codebase so they'll be good first issues as well.